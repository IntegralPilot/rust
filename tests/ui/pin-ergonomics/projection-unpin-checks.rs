//@ edition:2024
#![feature(pin_ergonomics, negative_impls)]
#![allow(incomplete_features)]

// This test verifies that a `&pin mut T` cannot be projected to a pinned
// reference field `&pin mut T.U` if neither `T: !Unpin` nor `U: Unpin`.

struct Foo<T>(T);
struct NotUnpinFoo<T>(T, std::marker::PhantomPinned);
struct NegUnpinFoo<T>(T);

impl<T> !Unpin for NegUnpinFoo<T> {}

fn unpin2not_unpin<T>(foo_mut: &pin mut Foo<T>, foo_const: &pin const Foo<T>) {
    let Foo(_x) = foo_mut; //~ ERROR the trait bound `Foo<T>: !Unpin` is not satisfied [E0277]
    let Foo(_x) = foo_const; //~ ERROR the trait bound `Foo<T>: !Unpin` is not satisfied [E0277]
    let Foo(_x) = (|| foo_mut)(); //~ ERROR the trait bound `Foo<T>: !Unpin` is not satisfied [E0277]
    let Foo(_x) = (|| foo_const)(); //~ ERROR the trait bound `Foo<T>: !Unpin` is not satisfied [E0277]
    let &mut _foo_mut = foo_mut; //~ ERROR mismatched types
    let &_foo_const = foo_const; //~ ERROR mismatched types
    // FIXME(pin_ergonomics): add these tests since `&pin` pattern is ready
    // let &pin mut _foo_mut = foo_mut; // error
    // let &pin mut _foo_const = foo_const; // error
}

fn unpin2unpin<T: Unpin>(foo_mut: &pin mut Foo<T>, foo_const: &pin const Foo<T>) {
    let Foo(_x) = foo_mut; // ok
    let Foo(_x) = foo_const; // ok
    let Foo(_x) = (|| foo_mut)(); // ok
    let Foo(_x) = (|| foo_const)(); // ok
    let &mut _foo_mut = foo_mut; //~ ERROR mismatched types
    let &_foo_const = foo_const; //~ ERROR mismatched types
    // FIXME(pin_ergonomics): add these tests since `&pin` pattern is ready
    // let &pin mut _foo_mut = foo_mut; // ok
    // let &pin mut _foo_const = foo_const; // ok
}

fn not_unpin2not_unpin<T>(foo_mut: &pin mut NotUnpinFoo<T>, foo_const: &pin const NotUnpinFoo<T>) {
    let NotUnpinFoo(_x, _) = foo_mut; //~ ERROR the trait bound `NotUnpinFoo<T>: !Unpin` is not satisfied [E0277]
    let NotUnpinFoo(_x, _) = foo_const; //~ ERROR the trait bound `NotUnpinFoo<T>: !Unpin` is not satisfied [E0277]
    let NotUnpinFoo(_x, _) = (|| foo_mut)(); //~ ERROR the trait bound `NotUnpinFoo<T>: !Unpin` is not satisfied [E0277]
    let NotUnpinFoo(_x, _) = (|| foo_const)(); //~ ERROR the trait bound `NotUnpinFoo<T>: !Unpin` is not satisfied [E0277]
    let &mut _foo_mut = foo_mut; //~ ERROR mismatched types
    let &_foo_const = foo_const; //~ ERROR mismatched types
    // FIXME(pin_ergonomics): add these tests since `&pin` pattern is ready
    // let &pin mut _foo_mut = foo_mut; // error
    // let &pin mut _foo_const = foo_const; // error
}

fn not_unpin2unpin<T: Unpin>(
    foo_mut: &pin mut NotUnpinFoo<T>,
    foo_const: &pin const NotUnpinFoo<T>,
) {
    let NotUnpinFoo(_x, _) = foo_mut; // ok
    let NotUnpinFoo(_x, _) = foo_const; // ok
    let NotUnpinFoo(_x, _) = (|| foo_mut)(); // ok
    let NotUnpinFoo(_x, _) = (|| foo_const)(); // ok
    let &mut _foo_mut = foo_mut; //~ ERROR mismatched types
    let &_foo_const = foo_const; //~ ERROR mismatched types
    // FIXME(pin_ergonomics): add these tests since `&pin` pattern is ready
    // let &pin mut _foo_mut = foo_mut; // ok
    // let &pin mut _foo_const = foo_const; // ok
}

fn neg_unpin2not_unpin<T>(foo_mut: &pin mut NegUnpinFoo<T>, foo_const: &pin const NegUnpinFoo<T>) {
    let NegUnpinFoo(_x) = foo_mut; // ok
    let NegUnpinFoo(_x) = foo_const; // ok
    let NegUnpinFoo(_x) = (|| foo_mut)(); // ok
    let NegUnpinFoo(_x) = (|| foo_const)(); // ok
    let &mut _foo_mut = foo_mut; //~ ERROR mismatched types
    let &_foo_const = foo_const; //~ ERROR mismatched types
    // FIXME(pin_ergonomics): add these tests since `&pin` pattern is ready
    // let &pin mut _foo_mut = foo_mut; // ok
    // let &pin mut _foo_const = foo_const; // ok
}

fn neg_unpin2unpin<T: Unpin>(
    foo_mut: &pin mut NegUnpinFoo<T>,
    foo_const: &pin const NegUnpinFoo<T>,
) {
    let NegUnpinFoo(_x) = foo_mut; // ok
    let NegUnpinFoo(_x) = foo_const; // ok
    let NegUnpinFoo(_x) = (|| foo_mut)(); // ok
    let NegUnpinFoo(_x) = (|| foo_const)(); // ok
    let &mut _foo_mut = foo_mut; //~ ERROR mismatched types
    let &_foo_const = foo_const; //~ ERROR mismatched types
    // FIXME(pin_ergonomics): add these tests since `&pin` pattern is ready
    // let &pin mut _foo_mut = foo_mut; // ok
    // let &pin mut _foo_const = foo_const; // ok
}

fn tuple_ref_mut_pat_and_pin_mut_of_tuple_mut_ty<'a, T>(
    (&mut x,): &'a pin mut (&'a mut NegUnpinFoo<T>,),
    //~^ ERROR the trait bound `&'a mut NegUnpinFoo<T>: !Unpin` is not satisfied in `(&'a mut NegUnpinFoo<T>,)` [E0277]
) -> &'a pin mut NegUnpinFoo<T> {
    x
}

fn tuple_ref_mut_pat_and_pin_mut_of_mut_tuple_ty<'a, T>(
    (&mut x,): &'a pin mut &'a mut (NegUnpinFoo<T>,),
    //~^ ERROR mismatched type
    //~| ERROR the trait bound `&'a mut (NegUnpinFoo<T>,): !Unpin` is not satisfied [E0277]
) -> &'a pin mut NegUnpinFoo<T> {
    x
}

fn ref_mut_tuple_pat_and_pin_mut_of_tuple_mut_ty<'a, T>(
    &mut (x,): &'a pin mut (&'a mut NegUnpinFoo<T>,), //~ ERROR mismatched type
) -> &'a pin mut NegUnpinFoo<T> {
    x
}

fn ref_mut_tuple_pat_and_pin_mut_of_mut_tuple_ty<'a, T>(
    &mut (x,): &'a pin mut &'a mut (NegUnpinFoo<T>,), //~ ERROR mismatched type
) -> &'a pin mut NegUnpinFoo<T> {
    x
}

fn tuple_pat_and_pin_mut_of_tuple_mut_ty<'a, T>(
    (x,): &'a pin mut (&'a mut NegUnpinFoo<T>,),
) -> &'a pin mut &'a mut NegUnpinFoo<T> {
    x // ok
}

fn tuple_pat_and_pin_mut_of_mut_tuple_ty<'a, T>(
    (x,): &'a pin mut &'a mut (NegUnpinFoo<T>,),
    //~^ ERROR the trait bound `&'a mut (NegUnpinFoo<T>,): !Unpin` is not satisfied [E0277]
) -> &'a pin mut NegUnpinFoo<T> {
    x
}

fn main() {}
