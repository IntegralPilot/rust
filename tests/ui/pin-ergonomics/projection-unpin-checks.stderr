error[E0277]: the trait bound `Foo<T>: !Unpin` is not satisfied
  --> $DIR/projection-unpin-checks.rs:15:9
   |
LL |     let Foo(_x) = foo_mut;
   |         ^^^^^^^ the trait bound `Foo<T>: !Unpin` is not satisfied

error[E0277]: the trait bound `Foo<T>: !Unpin` is not satisfied
  --> $DIR/projection-unpin-checks.rs:16:9
   |
LL |     let Foo(_x) = foo_const;
   |         ^^^^^^^ the trait bound `Foo<T>: !Unpin` is not satisfied

error[E0277]: the trait bound `Foo<T>: !Unpin` is not satisfied
  --> $DIR/projection-unpin-checks.rs:17:9
   |
LL |     let Foo(_x) = (|| foo_mut)();
   |         ^^^^^^^ the trait bound `Foo<T>: !Unpin` is not satisfied

error[E0277]: the trait bound `Foo<T>: !Unpin` is not satisfied
  --> $DIR/projection-unpin-checks.rs:18:9
   |
LL |     let Foo(_x) = (|| foo_const)();
   |         ^^^^^^^ the trait bound `Foo<T>: !Unpin` is not satisfied

error[E0308]: mismatched types
  --> $DIR/projection-unpin-checks.rs:19:9
   |
LL |     let &mut _foo_mut = foo_mut;
   |         ^^^^^^^^^^^^^   ------- this expression has type `Pin<&mut Foo<T>>`
   |         |
   |         expected `Pin<&mut Foo<T>>`, found `&mut _`
   |
   = note:         expected struct `Pin<&mut Foo<T>>`
           found mutable reference `&mut _`
help: you might have meant to use field `pointer` whose type is `&mut Foo<T>`
   |
LL |     let &mut _foo_mut = foo_mut.pointer;
   |                                ++++++++
help: to declare a mutable variable use
   |
LL -     let &mut _foo_mut = foo_mut;
LL +     let mut _foo_mut = foo_mut;
   |

error[E0308]: mismatched types
  --> $DIR/projection-unpin-checks.rs:20:9
   |
LL |     let &_foo_const = foo_const;
   |         ^^^^^^^^^^^   --------- this expression has type `Pin<&Foo<T>>`
   |         |
   |         expected `Pin<&Foo<T>>`, found `&_`
   |
   = note: expected struct `Pin<&Foo<T>>`
           found reference `&_`
help: you might have meant to use field `pointer` whose type is `&Foo<T>`
   |
LL |     let &_foo_const = foo_const.pointer;
   |                                ++++++++

error[E0308]: mismatched types
  --> $DIR/projection-unpin-checks.rs:31:9
   |
LL |     let &mut _foo_mut = foo_mut;
   |         ^^^^^^^^^^^^^   ------- this expression has type `Pin<&mut Foo<T>>`
   |         |
   |         expected `Pin<&mut Foo<T>>`, found `&mut _`
   |
   = note:         expected struct `Pin<&mut Foo<T>>`
           found mutable reference `&mut _`
help: you might have meant to use field `pointer` whose type is `&mut Foo<T>`
   |
LL |     let &mut _foo_mut = foo_mut.pointer;
   |                                ++++++++
help: to declare a mutable variable use
   |
LL -     let &mut _foo_mut = foo_mut;
LL +     let mut _foo_mut = foo_mut;
   |

error[E0308]: mismatched types
  --> $DIR/projection-unpin-checks.rs:32:9
   |
LL |     let &_foo_const = foo_const;
   |         ^^^^^^^^^^^   --------- this expression has type `Pin<&Foo<T>>`
   |         |
   |         expected `Pin<&Foo<T>>`, found `&_`
   |
   = note: expected struct `Pin<&Foo<T>>`
           found reference `&_`
help: you might have meant to use field `pointer` whose type is `&Foo<T>`
   |
LL |     let &_foo_const = foo_const.pointer;
   |                                ++++++++

error[E0277]: the trait bound `NotUnpinFoo<T>: !Unpin` is not satisfied
  --> $DIR/projection-unpin-checks.rs:39:9
   |
LL |     let NotUnpinFoo(_x, _) = foo_mut;
   |         ^^^^^^^^^^^^^^^^^^ the trait bound `NotUnpinFoo<T>: !Unpin` is not satisfied

error[E0277]: the trait bound `NotUnpinFoo<T>: !Unpin` is not satisfied
  --> $DIR/projection-unpin-checks.rs:40:9
   |
LL |     let NotUnpinFoo(_x, _) = foo_const;
   |         ^^^^^^^^^^^^^^^^^^ the trait bound `NotUnpinFoo<T>: !Unpin` is not satisfied

error[E0277]: the trait bound `NotUnpinFoo<T>: !Unpin` is not satisfied
  --> $DIR/projection-unpin-checks.rs:41:9
   |
LL |     let NotUnpinFoo(_x, _) = (|| foo_mut)();
   |         ^^^^^^^^^^^^^^^^^^ the trait bound `NotUnpinFoo<T>: !Unpin` is not satisfied

error[E0277]: the trait bound `NotUnpinFoo<T>: !Unpin` is not satisfied
  --> $DIR/projection-unpin-checks.rs:42:9
   |
LL |     let NotUnpinFoo(_x, _) = (|| foo_const)();
   |         ^^^^^^^^^^^^^^^^^^ the trait bound `NotUnpinFoo<T>: !Unpin` is not satisfied

error[E0308]: mismatched types
  --> $DIR/projection-unpin-checks.rs:43:9
   |
LL |     let &mut _foo_mut = foo_mut;
   |         ^^^^^^^^^^^^^   ------- this expression has type `Pin<&mut NotUnpinFoo<T>>`
   |         |
   |         expected `Pin<&mut NotUnpinFoo<T>>`, found `&mut _`
   |
   = note:         expected struct `Pin<&mut NotUnpinFoo<T>>`
           found mutable reference `&mut _`
help: you might have meant to use field `pointer` whose type is `&mut NotUnpinFoo<T>`
   |
LL |     let &mut _foo_mut = foo_mut.pointer;
   |                                ++++++++
help: to declare a mutable variable use
   |
LL -     let &mut _foo_mut = foo_mut;
LL +     let mut _foo_mut = foo_mut;
   |

error[E0308]: mismatched types
  --> $DIR/projection-unpin-checks.rs:44:9
   |
LL |     let &_foo_const = foo_const;
   |         ^^^^^^^^^^^   --------- this expression has type `Pin<&NotUnpinFoo<T>>`
   |         |
   |         expected `Pin<&NotUnpinFoo<T>>`, found `&_`
   |
   = note: expected struct `Pin<&NotUnpinFoo<T>>`
           found reference `&_`
help: you might have meant to use field `pointer` whose type is `&NotUnpinFoo<T>`
   |
LL |     let &_foo_const = foo_const.pointer;
   |                                ++++++++

error[E0308]: mismatched types
  --> $DIR/projection-unpin-checks.rs:58:9
   |
LL |     let &mut _foo_mut = foo_mut;
   |         ^^^^^^^^^^^^^   ------- this expression has type `Pin<&mut NotUnpinFoo<T>>`
   |         |
   |         expected `Pin<&mut NotUnpinFoo<T>>`, found `&mut _`
   |
   = note:         expected struct `Pin<&mut NotUnpinFoo<T>>`
           found mutable reference `&mut _`
help: you might have meant to use field `pointer` whose type is `&mut NotUnpinFoo<T>`
   |
LL |     let &mut _foo_mut = foo_mut.pointer;
   |                                ++++++++
help: to declare a mutable variable use
   |
LL -     let &mut _foo_mut = foo_mut;
LL +     let mut _foo_mut = foo_mut;
   |

error[E0308]: mismatched types
  --> $DIR/projection-unpin-checks.rs:59:9
   |
LL |     let &_foo_const = foo_const;
   |         ^^^^^^^^^^^   --------- this expression has type `Pin<&NotUnpinFoo<T>>`
   |         |
   |         expected `Pin<&NotUnpinFoo<T>>`, found `&_`
   |
   = note: expected struct `Pin<&NotUnpinFoo<T>>`
           found reference `&_`
help: you might have meant to use field `pointer` whose type is `&NotUnpinFoo<T>`
   |
LL |     let &_foo_const = foo_const.pointer;
   |                                ++++++++

error[E0308]: mismatched types
  --> $DIR/projection-unpin-checks.rs:70:9
   |
LL |     let &mut _foo_mut = foo_mut;
   |         ^^^^^^^^^^^^^   ------- this expression has type `Pin<&mut NegUnpinFoo<T>>`
   |         |
   |         expected `Pin<&mut NegUnpinFoo<T>>`, found `&mut _`
   |
   = note:         expected struct `Pin<&mut NegUnpinFoo<T>>`
           found mutable reference `&mut _`
help: you might have meant to use field `pointer` whose type is `&mut NegUnpinFoo<T>`
   |
LL |     let &mut _foo_mut = foo_mut.pointer;
   |                                ++++++++
help: to declare a mutable variable use
   |
LL -     let &mut _foo_mut = foo_mut;
LL +     let mut _foo_mut = foo_mut;
   |

error[E0308]: mismatched types
  --> $DIR/projection-unpin-checks.rs:71:9
   |
LL |     let &_foo_const = foo_const;
   |         ^^^^^^^^^^^   --------- this expression has type `Pin<&NegUnpinFoo<T>>`
   |         |
   |         expected `Pin<&NegUnpinFoo<T>>`, found `&_`
   |
   = note: expected struct `Pin<&NegUnpinFoo<T>>`
           found reference `&_`
help: you might have meant to use field `pointer` whose type is `&NegUnpinFoo<T>`
   |
LL |     let &_foo_const = foo_const.pointer;
   |                                ++++++++

error[E0308]: mismatched types
  --> $DIR/projection-unpin-checks.rs:85:9
   |
LL |     let &mut _foo_mut = foo_mut;
   |         ^^^^^^^^^^^^^   ------- this expression has type `Pin<&mut NegUnpinFoo<T>>`
   |         |
   |         expected `Pin<&mut NegUnpinFoo<T>>`, found `&mut _`
   |
   = note:         expected struct `Pin<&mut NegUnpinFoo<T>>`
           found mutable reference `&mut _`
help: you might have meant to use field `pointer` whose type is `&mut NegUnpinFoo<T>`
   |
LL |     let &mut _foo_mut = foo_mut.pointer;
   |                                ++++++++
help: to declare a mutable variable use
   |
LL -     let &mut _foo_mut = foo_mut;
LL +     let mut _foo_mut = foo_mut;
   |

error[E0308]: mismatched types
  --> $DIR/projection-unpin-checks.rs:86:9
   |
LL |     let &_foo_const = foo_const;
   |         ^^^^^^^^^^^   --------- this expression has type `Pin<&NegUnpinFoo<T>>`
   |         |
   |         expected `Pin<&NegUnpinFoo<T>>`, found `&_`
   |
   = note: expected struct `Pin<&NegUnpinFoo<T>>`
           found reference `&_`
help: you might have meant to use field `pointer` whose type is `&NegUnpinFoo<T>`
   |
LL |     let &_foo_const = foo_const.pointer;
   |                                ++++++++

error[E0277]: the trait bound `&'a mut NegUnpinFoo<T>: !Unpin` is not satisfied in `(&'a mut NegUnpinFoo<T>,)`
  --> $DIR/projection-unpin-checks.rs:93:5
   |
LL |     (&mut x,): &'a pin mut (&'a mut NegUnpinFoo<T>,),
   |     ^^^^^^^^^ within `(&'a mut NegUnpinFoo<T>,)`, the trait bound `&'a mut NegUnpinFoo<T>: !Unpin` is not satisfied
   |
   = note: required because it appears within the type `(&'a mut NegUnpinFoo<T>,)`

error[E0308]: mismatched types
  --> $DIR/projection-unpin-checks.rs:100:6
   |
LL |     (&mut x,): &'a pin mut &'a mut (NegUnpinFoo<T>,),
   |      ^^^^^^    ------------------------------------- expected due to this
   |      |
   |      expected `NegUnpinFoo<T>`, found `&mut _`
   |
   = note:         expected struct `NegUnpinFoo<T>`
           found mutable reference `&mut _`
note: to declare a mutable binding use: `mut x`
  --> $DIR/projection-unpin-checks.rs:100:6
   |
LL |     (&mut x,): &'a pin mut &'a mut (NegUnpinFoo<T>,),
   |      ^^^^^^
help: consider removing `&mut` from the pattern
   |
LL -     (&mut x,): &'a pin mut &'a mut (NegUnpinFoo<T>,),
LL +     (x,): &'a pin mut &'a mut (NegUnpinFoo<T>,),
   |

error[E0277]: the trait bound `&'a mut (NegUnpinFoo<T>,): !Unpin` is not satisfied
  --> $DIR/projection-unpin-checks.rs:100:5
   |
LL |     (&mut x,): &'a pin mut &'a mut (NegUnpinFoo<T>,),
   |     ^^^^^^^^^ the trait bound `&'a mut (NegUnpinFoo<T>,): !Unpin` is not satisfied

error[E0308]: mismatched types
  --> $DIR/projection-unpin-checks.rs:108:5
   |
LL |     &mut (x,): &'a pin mut (&'a mut NegUnpinFoo<T>,),
   |     ^^^^^^^^^  ------------------------------------- expected due to this
   |     |
   |     expected `Pin<&mut (&mut NegUnpinFoo<T>,)>`, found `&mut _`
   |
   = note:         expected struct `Pin<&'a mut (&'a mut NegUnpinFoo<T>,)>`
           found mutable reference `&mut _`
help: you might have meant to use field `pointer` whose type is `&'a mut (&'a mut NegUnpinFoo<T>,)`
   |
LL |     &mut (x,): &'a pin mut (&'a mut NegUnpinFoo<T>,).pointer,
   |                                                     ++++++++

error[E0308]: mismatched types
  --> $DIR/projection-unpin-checks.rs:114:5
   |
LL |     &mut (x,): &'a pin mut &'a mut (NegUnpinFoo<T>,),
   |     ^^^^^^^^^  ------------------------------------- expected due to this
   |     |
   |     expected `Pin<&mut &mut (NegUnpinFoo<T>,)>`, found `&mut _`
   |
   = note:         expected struct `Pin<&'a mut &'a mut (NegUnpinFoo<T>,)>`
           found mutable reference `&mut _`
help: you might have meant to use field `pointer` whose type is `&'a mut &'a mut (NegUnpinFoo<T>,)`
   |
LL |     &mut (x,): &'a pin mut &'a mut (NegUnpinFoo<T>,).pointer,
   |                                                     ++++++++

error[E0277]: the trait bound `&'a mut (NegUnpinFoo<T>,): !Unpin` is not satisfied
  --> $DIR/projection-unpin-checks.rs:126:5
   |
LL |     (x,): &'a pin mut &'a mut (NegUnpinFoo<T>,),
   |     ^^^^ the trait bound `&'a mut (NegUnpinFoo<T>,): !Unpin` is not satisfied

error: aborting due to 26 previous errors

Some errors have detailed explanations: E0277, E0308.
For more information about an error, try `rustc --explain E0277`.
